lib: deprecate G_snprintf() and replace with C99 snprintf()

nilason opened the pull request:
Beginning with UCRT in Visual Studio 2015 and Windows 10, the behaviour of `snprintf()` (and `vsnprintf()`) is now C99 standard conformant ([source](https://learn.microsoft.com/en-us/cpp/c-runtime-library/reference/snprintf-snprintf-snprintf-l-snwprintf-snwprintf-l?view=msvc-140#remarks)).

Thus, the output is now always null-terminated. The justification for an internal fix for that is no longer needed.

I should add that the code base is already using snprintf() in a number of places.

echoix commented:
When you say deprecate, is it only in some docs, or by using super strict compiler flags we can make our builds in CI fail for it, to not introduce back some usages of it? Users compiling with normal flags would still be able to successfully build.

wenzeslaus commented:
Do I understand correctly that the following in Microsoft documentation is a typo?

> Also, `snprintf()` includes one more character in the output because it doesn't null-terminate the buffer.

nilason commented:
> Do I understand correctly that the following in Microsoft documentation is a typo?
> 
> > Also, `snprintf()` includes one more character in the output because it doesn't null-terminate the buffer.

That's most likely. Further down you can read "snprintf always stores a terminating NULL character, truncating the output if necessary.".

nilason commented:
> When you say deprecate, is it only in some docs, or by using super strict compiler flags we can make our builds in CI fail for it, to not introduce back some usages of it? Users compiling with normal flags would still be able to successfully build.

I was foremost thinking in documentation terms (it will end up listed [here](https://grass.osgeo.org/programming8/deprecated.html)).
It will be easier the day we comply to >= C23, with the new [`[[ deprecated ]]`](https://en.cppreference.com/w/c/language/attributes/deprecated) attribute. We may consider implementing a macro for clang and/or gcc for checking this on CI, that is out of scope for this PR though.

wenzeslaus commented on lib/gis/timestamp.c:
```diff
@@ -425,11 +425,11 @@ int G_has_vector_timestamp(const char *name, const char *layer,
     char ele[GNAME_MAX];
 
     if (layer != NULL)
-        G_snprintf(ele, GNAME_MAX, "%s_%s", GV_TIMESTAMP_ELEMENT, layer);
+        snprintf(ele, GNAME_MAX, "%s_%s", GV_TIMESTAMP_ELEMENT, layer);
```
Doesn't this miss include stdio.h?

nilason replied:
Nice catch, I thought I looked through them all. Updated.

nilason commented:
Among Addons only [r.sim.water.mp](https://github.com/OSGeo/grass-addons/blob/a0e37beaf0ce8a4cc6ec615b6d3d40e39f9d725f/src/raster/r.sim.water.mp/output.c#L37) make use of G_snprintf(), whereas [i.wavelet](https://github.com/OSGeo/grass-addons/blob/a0e37beaf0ce8a4cc6ec615b6d3d40e39f9d725f/src/imagery/i.wavelet/main.c#L203) is the only one using snprintf().

echoix reviewed (commented):
I'll resume what I understand from all this, to do like [rubberducking](https://en.m.wikipedia.org/wiki/Rubber_duck_debugging), but for my comprehension of what I'm approving:

From all the changed files, there are two patterns used plus two exceptions.
- For C files, stdio is imported if not imported yet, then G_as printer is replaced with snprintf.
- For C++ files, cstdio is imported if not imported yet, then G_snprintf is replaced with std::snprintf.

For the two exceptions:
- the file that used to host G_snprintf is changed to use snprintf
- raster3d/r3.out.bin/main.c: this is the only file to double check. It has a unique change (with respect to the other changes in this PR). It declares the character array for the output filename with a size of GNAME_MAX instead of being a char pointer. Further below, it checks if there is an answer to the output parameter. If so, it is used as is with an empty string extension appended. If the output parameter doesn't have an answer, the `name` is used instead, and appended with a `.bin` extension. The complete snprintf call, since it returns the size (characters) that the buffer would have taken if it was successful, is used to check if it is bigger than the size of the outfile buffer (that was defined in this PR to GNAME_MAX). An error message is shown if the name is too long, and addition of this PR.

Ah, the rubberducking worked, there's a situation that isn't handled here.
If we look at the cppreference page, https://en.cppreference.com/w/cpp/io/c/fprintf

> 4) Number of characters that would have been written for a sufficiently large buffer if successful (not including the terminating null character), or a negative value if an error occurred. Thus, the (null-terminated) output has been completely written if and only if the returned value is nonnegative and less than `buf_size`.

The new logic seems to miss what happens when snprintf returns a negative value.

If it wasn't for that last point, I was approving this PR.

nilason commented:
> * the file that used to host G_snprintf is changed to use snprintf

This is not correct, it still uses `vsnprintf()`, but I have removed the unnecessary fix (adding null-terminator), and added notes (and cleaned up `#iimports`).

> * raster3d/r3.out.bin/main.c: this is the only file to double check. It has a unique change

The change of G_snprintf to C99 snprintf, made the compiler aware of a bug. The size of a pointer was checked instead of the size of the array. I addressed this and added a check whether the output string was truncated. This is the common practice (when checked at all!). This is not the PR to check for all potential errors.

echoix commented:
> > * the file that used to host G_snprintf is changed to use snprintf
> 
> 
> 
> This is not correct, it still uses `vsnprintf()`, but I have removed the unnecessary fix (adding null-terminator), and added notes (and cleaned up `#iimports`).
> 

You're right, I missed the v when writing back from memory and at my first reading. 

> 
> > * raster3d/r3.out.bin/main.c: this is the only file to double check. It has a unique change
> 
> 
> 
> The change of G_snprintf to C99 snprintf, made the compiler aware of a bug. The size of a pointer was checked instead of the size of the array. I addressed this and added a check whether the output string was truncated. This is the common practice (when checked at all!). This is not the PR to check for all potential errors.
> 
> 

So you are saying that it doesn't need to be checked for negative return value at all then? Or you do not want it here?

echoix commented:
Was it almost ready?

nilason commented:
> Was it almost ready?

It _is_ ready.
